llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -6908,6 +6908,90 @@ static Instruction *foldReductionIdiom(ICmpInst &I,
   return nullptr;
 }
 
+/// Helper function for foldICmpPHIWithMinMax
+Instruction *InstCombinerImpl::foldICmpPHIWithMinMaxHelper(
+    PHINode &PN, Instruction &I, Value *Z, ICmpInst::Predicate Pred) {
+
+  if (!PN.hasOneUse())
+    return nullptr;
+
+  bool Changed = false;
+  auto IsCondKnownTrue = [](Value *Val) -> std::optional<bool> {
+    if (!Val)
+      return std::nullopt;
+    if (match(Val, m_One()))
+      return true;
+    if (match(Val, m_Zero()))
+      return false;
+    return std::nullopt;
+  };
+
+  ICmpInst::Predicate SwappedPred =
+      ICmpInst::getNonStrictPredicate(ICmpInst::getSwappedPredicate(Pred));
+  for (unsigned OpNum = 0; OpNum != PN.getNumIncomingValues(); ++OpNum) {
+    if (auto *MinMax = dyn_cast<MinMaxIntrinsic>(PN.getIncomingValue(OpNum))) {
+      if (Pred != MinMax->getPredicate())
+        continue;
+
+      Value *X = MinMax->getLHS();
+      Value *Y = MinMax->getRHS();
+
+      SimplifyQuery Q =
+          SQ.getWithInstruction(PN.getIncomingBlock(OpNum)->getTerminator());
+
+      auto CmpXZ = IsCondKnownTrue(simplifyICmpInst(SwappedPred, X, Z, Q));
+      auto CmpYZ = IsCondKnownTrue(simplifyICmpInst(SwappedPred, Y, Z, Q));
+
+      if (!CmpXZ.has_value() && !CmpYZ.has_value())
+        continue;
+
+      if (CmpXZ.has_value() && CmpYZ.has_value()) {
+        auto CmpXY = IsCondKnownTrue(simplifyICmpInst(SwappedPred, X, Y, Q));
+        if (!CmpXY.has_value())
+          continue;
```

Comment:
Do you need to continue here?
I would think you could just ignore if both have a value and do you simplification of `X`/`Z`.
if `Y`/`Z` relationship is known, that will simplify later.

Reply:
I see so we can revert to how it was implemented before. I assume some other pass will probably take care of it?

Reply:
The point is that irrelevant of the ordering of `X`/`Y`, if `X >= Z` then `min(X, Y) < Z` -> `Y < Z`. Likewise for max.

Reply:
Also I think if we have an ordering for `X`/`Y`, the `min`/`max` will have simplified.

Reply:
Apologies I think I've misunderstood your comment. I think you're suggesting to remove the code block
```
if (!CmpXY.has_value())
  continue;
```
and just use X/Z to simplify since Y/Z relationship will be simplified later. I'm just failing to see how the Y/Z relationship will be simplified